### PR TITLE
dmclock: initial commit of dmclock QoS library

### PR DESCRIPTION
**Pull request by ivancich**
This commit brings in the dmclock library, located in a separate repo, as a git subtree. It also contains a README.git-subtree to describe this technique and some of the pitfalls surrounding it. And it links the cmake files so the library and support can be built.

Make targets include "dmclock", the static library; "dmclock-tests", unit tests for the dmclock code and supporting data structures; and "dmclock-sims", simulations that can be used to test and understand the library.

### CODE REVIEW

**Comment by ivancich** [comment]
Hmm, the two commits created by the "git subtree pull ..." command are not signed, and there do not appear to be command line options to add a signature.

**Comment by liewegas** [comment]
This looks good to me... @jdurgin ?

**Comment by jdurgin** [comment]
Don't worry about the signed-off-by for the subtree commits.

The 'make check' error looks real though - it got a failure from not running the dmclock tests.

**Comment by ivancich** [comment]
I'll handle the make check issue.

**Comment by liewegas** [comment]
retest this please

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -836,6 +848,18 @@ endif(WITH_TESTS)
...
 add_subdirectory(compressor)
 add_subdirectory(tools)
+add_subdirectory(dmclock) # after gmock
+
+add_dependencies(tests
+  dmclock-tests dmclock-data-struct-tests)
+
+if(WITH_TESTS)
+install(PROGRAMS
+  ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/dmclock-tests
```
Comment (tchaikov) [inline src/CMakeLists.txt]: might want to include this in "ceph-test". see debian/ceph-test.install and ceph.spec.in (search for `%files -n ceph-test`)

Comment (ivancich) [inline src/CMakeLists.txt]: Will do.

**Comment by liewegas** [comment]
-- Set runtime path of "/tmp/buildd/ceph-12.0.1-1259-g49dfa5a/debian/tmp/usr/bin/ceph-mds" to "/usr/lib/ceph"
CMake Error at src/cmake_install.cmake:193 (file):
  file INSTALL cannot find
  "/tmp/buildd/ceph-12.0.1-1259-g49dfa5a/obj-x86_64-linux-gnu/bin/dmclock-tests".
Call Stack (most recent call first):
  cmake_install.cmake:37 (include)
  


https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=xenial,DIST=xenial,MACHINE_SIZE=huge/2765//consoleFull

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-01_20:13:55-rados-wip-sage-testing---basic-smithi/

**Comment by smithfarm** [comment]
This breaks SUSE builds:

```
[  158s] CMake Error at src/dmclock/CMakeLists.txt:20 (find_package):
[  158s]   By not providing "Findgtest.cmake" in CMAKE_MODULE_PATH this project has
[  158s]   asked CMake to find a package configuration file provided by "gtest", but
[  158s]   CMake did not find one.
[  158s] 
[  158s]   Could not find a package configuration file provided by "gtest" with any of
[  158s]   the following names:
[  158s] 
[  158s]     gtestConfig.cmake
[  158s]     gtest-config.cmake
[  158s] 
[  158s]   Add the installation prefix of "gtest" to CMAKE_PREFIX_PATH or set
[  158s]   "gtest_DIR" to a directory containing one of the above files.  If "gtest"
[  158s]   provides a separate development package or SDK, be sure it has been
[  158s]   installed.
[  158s] 
[  158s] 
[  158s] -- Configuring incomplete, errors occurred!
```

**Comment by ivancich** [comment]
Hi Nathan, Thanks for looking into that. When ceph is built, gtest and gtest_main are targets, and I thought that would have prevented the find_package of gtest. Would you mind looking into that?

Nonetheless, I'll fix the cases to match cmake's upstream. I'll likely include that in https://github.com/ceph/ceph/pull/14997 , which is currently under eval.

**Comment by smithfarm** [comment]
@ivancich in #14997 you are dropping (i.e. not packaging) the binaries `dmclock-tests` and `dmclock-data-struct-tests`, right? In that case, instead of `if(WITH_TESTS)` you could guard the build of these binaries with some other conditional that defaults to OFF? 

I ask because `WITH_TESTS=ON` means "the binaries in ceph-test are to be built". It does not mean "ctest is to be run". . .

**Comment by smithfarm** [comment]
@ivancich See https://github.com/ceph/ceph/pull/10872 for how/why of WITH_TESTS

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -842,6 +854,18 @@ add_subdirectory(compressor)
...
+
+add_subdirectory(dmclock) # after gmock
+add_dependencies(tests dmclock-tests dmclock-data-struct-tests)
+
+if(WITH_TESTS)
+  install(PROGRAMS
+    ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/dmclock-tests
+    ${CMAKE_RUNTIME_OUTPUT_DIRECTORY}/dmclock-data-struct-tests
+    DESTINATION bin)
+endif(WITH_TESTS)
```
Comment (smithfarm) [inline src/CMakeLists.txt]: @ivancich This is where the build of `dmclock-tests` and `dmclock-data-struct-tests` is made conditional upon `WITH_TESTS`.

Comment (smithfarm) [inline src/CMakeLists.txt]: If we could change the guard to some other conditional (and not package these two binaries), then the problem I'm having will be solved I think!

Comment (smithfarm) [inline src/CMakeLists.txt]: e.g. in `CMakeLists.txt` add a line that says `set(WITH_DMCLOCK_TESTS "enable the build of dmclock-tests and dmclock-data-struct-tests binaries` OFF)`

and then s/WITH_TESTS/WITH_DMCLOCK_TESTS/

Comment (smithfarm) [inline src/CMakeLists.txt]: Then somebody who wanted to build these binaries could pass `-DWITH_DMCLOCK_TESTS=ON` to cmake, but for purposes of building RPMs and debs we would use `-DWITH_DMCLOCK_TESTS=OFF`.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -842,6 +854,18 @@ add_subdirectory(compressor)
 
 add_subdirectory(tools)
 
+# dmClock
+
+add_subdirectory(dmclock) # after gmock
+add_dependencies(tests dmclock-tests dmclock-data-struct-tests)
```
Comment (smithfarm) [inline src/CMakeLists.txt]: this line would get moved into the proposed `WITH_DMCLOCK_TESTS` conditional

File: src/CMakeLists.txt (hunk 2 above)
Comment (ivancich) [inline src/CMakeLists.txt]: If you look at where this is heading (https://github.com/ivancich/ceph-fork/tree/wip-bring-in-dmclock-p2-gtest) all that is gone. dmclock tests won't even be built unless you do something like "make dmclock-tests". They're detached from ceph's tests. I think that'll give you what you want, no?

Comment (smithfarm) [inline src/CMakeLists.txt]: > dmclock tests won't even be built unless you do something like "make dmclock-tests". They're detached from ceph's tests. I think that'll give you what you want, no?

Yes.

Comment (smithfarm) [inline src/CMakeLists.txt]: although this [1] appears to be building the dmclock tests

[1] https://github.com/ivancich/ceph-fork/blob/wip-bring-in-dmclock-p2-gtest/src/dmclock/CMakeLists.txt#L19-L25

**Pull request merged**